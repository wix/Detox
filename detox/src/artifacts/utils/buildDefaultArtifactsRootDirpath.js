const path = require('path');

const getTimeStampString = require('./getTimeStampString');

function buildDefaultRootForArtifactsRootDirpath(configuration, rootDir) {
  if (rootDir.endsWith('/') || rootDir.endsWith('\\')) {
    return rootDir;
  }

  const seed = Number(process.env.DETOX_START_TIMESTAMP || String(Date.now()));
  const subdir = `${configuration}.${getTimeStampString(new Date(seed))}`;
  return path.join(rootDir, subdir);
}

module.exports = buildDefaultRootForArtifactsRootDirpath;
