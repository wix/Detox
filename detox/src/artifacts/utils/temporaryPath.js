const path = require('path');

const tempfile = require('tempfile');

module.exports = {
  for: {
    png: () => tempfile('.detox.png'),
    log: () => tempfile('.detox.log'),
    mp4: () => tempfile('.detox.mp4'),
    dtxrec: () => tempfile('.detox.dtxrec'),
    viewhierarchy: () => tempfile('.detox.viewhierarchy'),
  },
  mask: () => path.join(tempfile(), '..') + path.sep + '*.detox.*',
};
