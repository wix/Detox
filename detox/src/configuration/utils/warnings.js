const DEVICE_LAUNCH_ARGS_DEPRECATION = `\
--device-launch-args is a deprecated name of this option.
Please change your scripts to use --device-boot-args instead.`;

const DEVICE_LAUNCH_ARGS_GENERIC_DEPRECATION = `\
--device-launch-args / $DETOX_DEVICE_LAUNCH_ARGS is a deprecated name of this option.
Please change your scripts to use --device-boot-args / $DETOX_DEVICE_BOOT_ARGS instead.`;

module.exports = {
  DEVICE_LAUNCH_ARGS_DEPRECATION,
  DEVICE_LAUNCH_ARGS_GENERIC_DEPRECATION,
};
