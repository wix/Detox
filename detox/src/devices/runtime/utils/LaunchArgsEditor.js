const _ = require('lodash');

const log = require('../../../utils/logger');

const ScopedLaunchArgsEditor = require('./ScopedLaunchArgsEditor');

/**
 * @typedef {Object} LaunchArgsEditorOptions
 * @property {boolean} [permanent=false] - Indicates whether the operation should affect the permanent app launch args.
 */

class LaunchArgsEditor {
  constructor() {
    this._local = new ScopedLaunchArgsEditor();
    this._shared = new ScopedLaunchArgsEditor();
  }

  get shared() {
    return this._shared;
  }

  /**
   * @param {LaunchArgsEditorOptions} [options] - deprecated
   */
  modify(launchArgs, options) {
    this._assertNoDeprecatedOptions('modify', options);

    if (!_.isEmpty(launchArgs)) {
      if (options && options.permanent) {
        this._shared.modify(launchArgs);
      } else {
        this._local.modify(launchArgs);
      }
    }

    return this;
  }

  /**
   * @param {LaunchArgsEditorOptions} [options] - deprecated
   */
  reset(options) {
    this._assertNoDeprecatedOptions('reset', options);
    this._local.reset();

    if (options && options.permanent) {
      this._shared.reset();
    }

    return this;
  }

  /**
   * @param {LaunchArgsEditorOptions} [options] - deprecated
   */
  get(options) {
    this._assertNoDeprecatedOptions('get', options);
    const permanent = options && options.permanent;

    if (permanent === true) {
      return this._shared.get();
    }

    if (permanent === false) {
      return this._local.get();
    }

    return _.merge(this._shared.get(), this._local.get());
  }

  _assertNoDeprecatedOptions(methodName, options) {
    if (!options) {
      return;
    }

    const target = options.permanent ? 'appLaunchArgs.shared' : 'appLaunchArgs';
    log.warn({ event: 'DEPRECATION' }, [
      `Detected an attempt to use device.launchArgs.${methodName}(...) with a deprecated 'options' parameter.`,
      `Please use device.${target}.${methodName}(...) instead.`,
      `For more details, see the reference at: https://wix.github.io/Detox/docs/api/device-object-api#deviceapplaunchargs\n`,
    ].join('\n'));
  }
}

module.exports = LaunchArgsEditor;
