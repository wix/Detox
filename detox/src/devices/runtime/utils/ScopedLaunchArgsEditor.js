const Storage = require('./Storage');

class ScopedLaunchArgsEditor {
  constructor() {
    this._storage = new Storage();
  }

  get() {
    return this._storage.get();
  }

  reset() {
    this._storage.reset();
    return this;
  }

  modify(launchArgs) {
    this._storage.assign(launchArgs);
    return this;
  }
}

module.exports = ScopedLaunchArgsEditor;
