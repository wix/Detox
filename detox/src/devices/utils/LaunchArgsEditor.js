const _ = require('lodash');

const Storage = require('./Storage');

/**
 * @typedef {Object} LaunchArgsEditorOptions
 * @property {boolean} [permanent=false] - Indicates whether the operation should affect the permanent app launch args.
 */

class LaunchArgsEditor {
  constructor() {
    this._permanent = new Storage();
    this._transient = new Storage();
  }

  /** @param {LaunchArgsEditorOptions} [options] */
  modify(launchArgs, options) {
    if (!_.isEmpty(launchArgs)) {
      if (options && options.permanent) {
        this._permanent.assign(launchArgs);
      } else {
        this._transient.assign(launchArgs);
      }
    }

    return this;
  }

  /** @param {LaunchArgsEditorOptions} [options] */
  reset(options) {
    this._transient.reset();

    if (options && options.permanent) {
      this._permanent.reset();
    }

    return this;
  }

  /** @param {LaunchArgsEditorOptions} [options] */
  get(options) {
    const value = this._permanent.get();

    if (options && options.permanent) {
      return value;
    } else {
      return _.merge(value, this._transient.get());
    }
  }
}

module.exports = LaunchArgsEditor;
