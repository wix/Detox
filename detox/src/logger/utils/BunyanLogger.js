const { PassThrough } = require('stream');
const { promisify } = require('util');

const bunyan = require('bunyan');
const bds = require('bunyan-debug-stream');
const _ = require('lodash');

const { DetoxInternalError } = require('../../errors');

class BunyanLogger {
  constructor() {
    this._bunyan = bunyan.createLogger({ name: 'detox', streams: [] });
    /** @type {bunyan.Stream} */
    this._debugStream = null;
    /** @type {bunyan.Stream} */
    this._fileStream = null;
  }

  /**
   * @returns {bunyan}
   */
  get logger() {
    return this._bunyan;
  }

  /**
   * @param {Detox.DetoxLoggerConfig} config
   * @returns {this}
   */
  installDebugStream(config) {
    if (this._debugStream) {
      _.remove(this._bunyan['streams'], this._debugStream);
      // @ts-ignore
      this._debugStream.stream.end();
      this._debugStream = null;
    }

    const streamOptions = { out: null, ...config.options };
    /* istanbul ignore next */
    if (!streamOptions.out) {
      // This is a default if-else branch, used everywhere except for the unit tests.
      streamOptions.out = new PassThrough().pipe(process.stderr);
    }

    this._bunyan.addStream({
      type: 'raw',
      level: config.level,
      stream: bds.default(streamOptions),
    });

    this._debugStream = _.last(this._bunyan['streams']);

    return this;
  }

  /**
   * @param {string} file
   * @returns {this}
   */
  installFileStream(file) {
    /* istanbul ignore next */
    if (this._fileStream) {
      // This is an impossible condition, but we keep it here for the sake of completeness.
      throw new DetoxInternalError('Trying to install a second file stream inside already initialized Bunyan logger');
    }

    this._fileStream = {
      level: 'trace',
      path: file,
    };

    this._bunyan.addStream(this._fileStream);

    return this;
  }

  async closeFileStreams() {
    const internalBunyanStreams = this._bunyan['streams'];
    const openFileStreams = _.filter(internalBunyanStreams, this._isOpenFileStream);
    _.remove(internalBunyanStreams, openFileStreams);

    await Promise.all(openFileStreams.map(bunyanStream => {
      const stream = bunyanStream.stream;
      return promisify(stream.end.bind(stream))();
    }));
  }

  /** @private */
  _isOpenFileStream = (bunyanStream) => {
    switch (bunyanStream.type) {
      case 'file':
        return bunyanStream.path && !bunyanStream.stream.destroyed;
      case 'raw':
        /* istanbul ignore next */
        const stream = bunyanStream.stream === this._debugStream.stream
          ? bunyanStream.stream._out
          : bunyanStream.stream;

        /* istanbul ignore next */
        return stream.fd > 2 && !stream.closed;
    }
  };
}

module.exports = BunyanLogger;
