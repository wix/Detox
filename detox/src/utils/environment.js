const fs = require('fs');
const os = require('os');
const path = require('path');

const exec = require('child-process-promise').exec;
const ini = require('ini');
const _ = require('lodash');
const _which = require('which');

const DetoxRuntimeError = require('../errors/DetoxRuntimeError');

const appdatapath = require('./appdatapath');
const fsext = require('./fsext');

function which(executable, path) {
  return _which.sync(executable, { path, nothrow: true });
}

const DETOX_LIBRARY_ROOT_PATH = path.join(appdatapath.appDataPath(), 'Detox');
const MISSING_SDK_ERROR = `$ANDROID_SDK_ROOT is not defined, set the path to the SDK installation directory into $ANDROID_SDK_ROOT,
Go to https://developer.android.com/studio/command-line/variables.html for more details`;
const DEVICE_LOCK_FILE_PATH_IOS = path.join(DETOX_LIBRARY_ROOT_PATH, 'device.registry.state.lock');
const DEVICE_LOCK_FILE_PATH_ANDROID = path.join(DETOX_LIBRARY_ROOT_PATH, 'android-device.registry.state.lock');
const GENYCLOUD_GLOBAL_CLEANUP_FILE_PATH = path.join(DETOX_LIBRARY_ROOT_PATH, 'genycloud-cleanup.lock');
const LAST_FAILED_TESTS_PATH = path.join(DETOX_LIBRARY_ROOT_PATH, 'last-failed.txt');

function getAndroidSDKPath() {
  return process.env.ANDROID_SDK_ROOT || process.env.ANDROID_HOME || '';
}

function getAndroidSDKHome() {
  return process.env['ANDROID_SDK_HOME'] || os.homedir();
}

function getEmulatorHome() {
  return process.env['ANDROID_EMULATOR_HOME'] || path.join(getAndroidSDKHome(), '.android');
}

function getAvdHome() {
  return process.env['ANDROID_AVD_HOME'] || path.join(getEmulatorHome(), 'avd');
}

function getAvdDir(avdName) {
  const avdIniPath = path.join(getAvdHome(), `${avdName}.ini`);
  if (!fs.existsSync(avdIniPath)) {
    throwMissingAvdINIError(avdName, avdIniPath);
  }

  const avdIni = ini.parse(fs.readFileSync(avdIniPath, 'utf-8'));
  if (!fs.existsSync(avdIni.path)) {
    throwMissingAvdError(avdName, avdIni.path, avdIniPath);
  }

  return avdIni.path;
}

function getAvdManagerPath() {
  return path.join(getAndroidSDKPath(), 'cmdline-tools', 'latest', 'bin', 'avdmanager');
}

function getAndroidSdkManagerPath() {
  return path.join(getAndroidSDKPath(), 'cmdline-tools', 'latest', 'bin', 'sdkmanager');
}

function getAndroidEmulatorPath() {
  const sdkRoot = getAndroidSDKPath();
  if (!sdkRoot) {
    return which('emulator') || throwMissingSdkError();
  }

  const defaultPath = which('emulator', path.join(sdkRoot, 'emulator'));
  if (defaultPath) {
    return defaultPath;
  }

  const legacyPath = which('emulator', path.join(sdkRoot, 'tools'));
  if (legacyPath) {
    return legacyPath;
  }

  throwSdkIntegrityError(sdkRoot, 'emulator/emulator');
}

async function getAaptPath() {
  const sdkRoot = getAndroidSDKPath();
  if (!sdkRoot) {
    return which('aapt') || throwMissingSdkError();
  }

  const latestBuildTools = await getLatestBuildToolsPath(sdkRoot);
  const defaultPath = latestBuildTools && which('aapt', latestBuildTools);
  if (defaultPath) {
    return defaultPath;
  }

  throwSdkIntegrityError(sdkRoot, `${latestBuildTools}/aapt`);
}

async function getLatestBuildToolsPath(sdkRoot) {
  if (!sdkRoot) return '';

  const buildToolsDir = path.join(sdkRoot, 'build-tools');
  if (!fs.existsSync(buildToolsDir)) return '';

  const buildToolsVersions = await fsext.getDirectories(buildToolsDir);
  const latestBuildToolsVersion = _.last(buildToolsVersions);
  if (!latestBuildToolsVersion) return '';

  return path.join(buildToolsDir, latestBuildToolsVersion);
}

function getAdbPath() {
  const sdkRoot = getAndroidSDKPath();
  if (!sdkRoot) {
    return which('adb') || throwMissingSdkError();
  }

  const defaultPath = which('adb', path.join(sdkRoot, 'platform-tools'));
  if (defaultPath) {
    return defaultPath;
  }

  throwSdkIntegrityError(sdkRoot, 'platform-tools/adb');
}

function getGmsaasPath() {
  return which('gmsaas') || throwMissingGmsaasError();
}

function throwMissingSdkError() {
  throw new DetoxRuntimeError(MISSING_SDK_ERROR);
}

function throwMissingAvdINIError(avdName, avdIniPath) {
  throw new DetoxRuntimeError(`Failed to find INI file for ${avdName} at path: ${avdIniPath}`);
}

function throwMissingAvdError(avdName, avdPath, avdIniPath) {
  throw new DetoxRuntimeError(
    `Failed to find AVD ${avdName} directory at path: ${avdPath}\n` +
    `Please verify "path" property in the INI file: ${avdIniPath}`
  );
}

function throwSdkIntegrityError(sdkRoot, relativeExecutablePath) {
  const executablePath = path.join(sdkRoot, relativeExecutablePath);
  const name = path.basename(executablePath);
  const dir = path.dirname(executablePath);

  throw new DetoxRuntimeError(
    `There was no "${name}" executable file in directory: ${dir}.\n` +
    `Check integrity of your Android SDK.`
  );
}

function throwMissingGmsaasError() {
  throw new DetoxRuntimeError(`Failed to locate Genymotion's gmsaas executable. Please add it to your $PATH variable!\nPATH is currently set to: ${process.env.PATH}`);
}

function getDetoxVersion() {
  return require(path.join(__dirname, '../../package.json')).version;
}

let _iosFrameworkPath;
async function getFrameworkPath() {
  if (!_iosFrameworkPath) {
    _iosFrameworkPath = _doGetFrameworkPath();
  }

  return _iosFrameworkPath;
}

async function _doGetFrameworkPath() {
  const detoxVersion = getDetoxVersion();
  const sha1 = (await exec(`(echo "${detoxVersion}" && xcodebuild -version) | shasum | awk '{print $1}'`)).stdout.trim();
  return `${DETOX_LIBRARY_ROOT_PATH}/ios/${sha1}/Detox.framework`;
}

function getDetoxLibraryRootPath() {
  return DETOX_LIBRARY_ROOT_PATH;
}

function getDeviceLockFilePathIOS() {
  return DEVICE_LOCK_FILE_PATH_IOS;
}

// TODO This can probably be merged with IOS' by now
function getDeviceLockFilePathAndroid() {
  return DEVICE_LOCK_FILE_PATH_ANDROID;
}

function getGenyCloudGlobalCleanupFilePath() {
  return GENYCLOUD_GLOBAL_CLEANUP_FILE_PATH;
}

function getLastFailedTestsPath() {
  return LAST_FAILED_TESTS_PATH;
}

function getHomeDir() {
  return os.homedir();
}

module.exports = {
  getAaptPath,
  getAdbPath,
  getAvdHome,
  getAvdDir,
  getAvdManagerPath,
  getAndroidSdkManagerPath,
  getGmsaasPath,
  getDetoxVersion,
  getFrameworkPath,
  getAndroidSDKPath,
  getAndroidEmulatorPath,
  getDetoxLibraryRootPath,
  getDeviceLockFilePathIOS,
  getDeviceLockFilePathAndroid,
  getGenyCloudGlobalCleanupFilePath,
  getLastFailedTestsPath,
  getHomeDir,
};
