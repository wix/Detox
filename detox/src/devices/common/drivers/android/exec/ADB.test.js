describe('ADB', () => {
  const deviceId = 'mockEmulator';
  const adbBinPath = `/Android/sdk-mock/platform-tools/adb`;

  let ADB;
  let adb;
  let DeviceHandle;
  let EmulatorHandle;
  let exec;
  let spawn;

  beforeEach(() => {
    jest.mock('../../../../../utils/logger');
    jest.mock('../../../../../utils/environment');
    require('../../../../../utils/environment').getAdbPath.mockReturnValue(adbBinPath);

    jest.mock('../../../../../utils/encoding', () => ({
      encodeBase64: (text) => `base64(${text})`,
    }));

    jest.mock('../tools/DeviceHandle');
    DeviceHandle = require('../tools/DeviceHandle');

    jest.mock('../tools/EmulatorHandle');
    EmulatorHandle = require('../tools/EmulatorHandle');

    jest.mock('../../../../../utils/exec', () => ({
      execWithRetriesAndLogs: jest.fn().mockReturnValue({ stdout: '' }),
      spawnAndLog: jest.fn(),
    }));
    exec = require('../../../../../utils/exec').execWithRetriesAndLogs;
    spawn = require('../../../../../utils/exec').spawnAndLog;

    ADB = require('./ADB');
    adb = new ADB();
  });

  describe('devices', () => {
    const mockDevices = [
      'MOCK_SERIAL\tdevice',
      '192.168.60.101:6666\tdevice',
      'emulator-5554\tdevice',
      'emulator-5556\toffline',
    ];
    const adbDevices = ['List of devices attached', ...mockDevices, ''].join('\n');

    it(`should invoke ADB`, async () => {
      await adb.devices();
      expect(exec).toHaveBeenCalledWith(`"${adbBinPath}"  devices`, { verbosity: 'high', retries: 1 });
      expect(exec).toHaveBeenCalledTimes(1);
    });

    it('should return proper, type-based device handles', async () => {
      exec.mockReturnValue({ stdout: adbDevices });

      const { devices, stdout } = await adb.devices();

      expect(stdout).toBe(adbDevices);
      expect(devices).toHaveLength(4);
      expect(devices).toEqual([
        DeviceHandle.mock.instances[0],
        DeviceHandle.mock.instances[1],
        EmulatorHandle.mock.instances[0],
        EmulatorHandle.mock.instances[1],
      ]);
      expect(DeviceHandle).toHaveBeenCalledWith(mockDevices[0]);
      expect(DeviceHandle).toHaveBeenCalledWith(mockDevices[1]);
      expect(EmulatorHandle).toHaveBeenCalledWith(mockDevices[2]);
      expect(EmulatorHandle).toHaveBeenCalledWith(mockDevices[3]);
    });

    it(`should return an empty list if no devices are available`, async () => {
      exec.mockReturnValue({
        stdout: 'List of devices attached\n'
      });

      const { devices } = await adb.devices();
      expect(devices.length).toEqual(0);
    });
  });

  it(`install`, async () => {
    await adb.install('emulator-5556', 'path inside "quotes" to/app');

    expect(exec).toHaveBeenCalledWith(
      expect.stringContaining('adb" -s emulator-5556 shell "getprop ro.build.version.sdk"'),
      { retries: 5 });
  });

  it(`install api 22`, async () => {
    jest.spyOn(adb, 'apiLevel').mockImplementation(async () => 22);
    await adb.install('emulator-5556', 'path inside "quotes" to/app');

    expect(exec).toHaveBeenCalledWith(
      expect.stringContaining('adb" -s emulator-5556 install -rg "path inside \\"quotes\\" to/app"'),
      { retries: 1 });
  });

  it(`install api 23`, async () => {
    jest.spyOn(adb, 'apiLevel').mockImplementation(async () => 23);
    await adb.install('emulator-5556', 'path inside "quotes" to/app');

    expect(exec).toHaveBeenCalledWith(
      expect.stringContaining('adb" -s emulator-5556 install -r -g -t "path inside \\"quotes\\" to/app"'),
      { retries: 1 });
  });

  it(`uninstall`, async () => {
    await adb.uninstall('com.package');
    expect(exec).toHaveBeenCalledTimes(1);
  });

  it(`terminate`, async () => {
    await adb.terminate('com.package');
    expect(exec).toHaveBeenCalledTimes(1);
  });

  it(`sets location both with commas and dots due to the locale issue`, async () => {
    const lat = 30.5;
    const lon = -70.5;

    await adb.setLocation(deviceId, lat, lon);

    expect(exec).toHaveBeenCalledWith(
      expect.stringContaining(`-s mockEmulator emu "geo fix -70.5 30.5"`),
      expect.anything());

    expect(exec).toHaveBeenCalledWith(
      expect.stringContaining(`-s mockEmulator emu "geo fix -70,5 30,5"`),
      expect.anything());
  });

  it(`pidof (success)`, async () => {
    jest.spyOn(adb, 'shell').mockImplementation(async () =>
      `u0_a19        2199  1701 3554600  70264 0                   0 s com.google.android.ext.services `);

    expect(await adb.pidof('', 'com.google.android.ext.services')).toBe(2199);
  });

  it(`pidof (failure)`, async () => {
    jest.spyOn(adb, 'shell').mockImplementation(async () => '');
    expect(await adb.pidof('', 'com.google.android.ext.services')).toBe(NaN);
  });

  it('push', async () => {
    const sourceFile = '/mock-source/file.xyz';
    const destFile = '/sdcard/file.abc';
    await adb.push(deviceId, sourceFile, destFile);

    expect(exec).toHaveBeenCalledWith(
      expect.stringContaining(`-s mockEmulator push "${sourceFile}" "${destFile}"`),
      expect.anything());
  });

  it('remote-install api 22', async () => {
    jest.spyOn(adb, 'apiLevel').mockImplementation(async () => 22);
    const binaryPath = '/mock-path/filename.mock';
    await adb.remoteInstall(deviceId, binaryPath);

    expect(exec).toHaveBeenCalledWith(
      expect.stringContaining(`-s mockEmulator shell "pm install -rg ${binaryPath}"`),
      expect.anything());
  });

  it('remote-install api 23', async () => {
    jest.spyOn(adb, 'apiLevel').mockImplementation(async () => 23);
    const binaryPath = '/mock-path/filename.mock';
    await adb.remoteInstall(deviceId, binaryPath);

    expect(exec).toHaveBeenCalledWith(
      expect.stringContaining(`-s mockEmulator shell "pm install -r -g -t ${binaryPath}"`),
      expect.anything());
  });

  it('global text-typing', async () => {
    const text = 'some-text-with spaces';
    const expectedText = 'some-text-with%sspaces';
    await adb.typeText(deviceId, text);
    expect(exec).toHaveBeenCalledWith(
      expect.stringContaining(`-s mockEmulator shell "input text ${expectedText}"`),
      expect.anything());
  });

  describe('unlockScreen', () => {
    async function unlockScreenWithPowerStatus(mWakefulness, mUserActivityTimeoutOverrideFromWindowManager) {
      jest.spyOn(adb, 'shell').mockImplementation(async () => `
        mWakefulness=${mWakefulness}
        mWakefulnessChanging=false
        mWakeLockSummary=0x0
        mUserActivitySummary=0x1
        mWakeUpWhenPluggedOrUnpluggedConfig=false
        mWakeUpWhenPluggedOrUnpluggedInTheaterModeConfig=false
        mUserActivityTimeoutOverrideFromWindowManager=${mUserActivityTimeoutOverrideFromWindowManager}
        mUserInactiveOverrideFromWindowManager=false
      `);

      await adb.unlockScreen(deviceId);
    }

    describe('when unlocking an awake and unlocked device', function() {
      beforeEach(async () => unlockScreenWithPowerStatus('Awake', '-1'));

      it('should not press power button', () =>
        expect(adb.shell).not.toHaveBeenCalledWith(deviceId, 'input keyevent KEYCODE_POWER'));

      it('should not press menu button', () =>
        expect(adb.shell).not.toHaveBeenCalledWith(deviceId, 'input keyevent KEYCODE_MENU'));
    });

    describe('when unlocking a sleeping and locked device', function() {
      beforeEach(async () => unlockScreenWithPowerStatus('Asleep', '10000'));

      it('should press power button first', () =>
        expect(adb.shell.mock.calls[1]).toEqual([deviceId, 'input keyevent KEYCODE_POWER']));

      it('should press menu afterwards', () =>
        expect(adb.shell.mock.calls[2]).toEqual([deviceId, 'input keyevent KEYCODE_MENU']));
    });

    describe('when unlocking an awake but locked device', function() {
      beforeEach(async () => unlockScreenWithPowerStatus('Awake', '10000'));

      it('should not press power button', () =>
        expect(adb.shell).not.toHaveBeenCalledWith(deviceId, 'input keyevent KEYCODE_POWER'));

      it('should press menu button', () =>
        expect(adb.shell).toHaveBeenCalledWith(deviceId, 'input keyevent KEYCODE_MENU'));
    });

    describe('when unlocking a sleeping but unlocked device', function() {
      beforeEach(async () => unlockScreenWithPowerStatus('Asleep', '-1'));

      it('should press power button', () =>
        expect(adb.shell).toHaveBeenCalledWith(deviceId, 'input keyevent KEYCODE_POWER'));

      it('should not press menu button', () =>
        expect(adb.shell).not.toHaveBeenCalledWith(deviceId, 'input keyevent KEYCODE_MENU'));
    });
  });

  describe('spawnInstrumentation', () => {
    const testRunner = 'aTestRunner';

    it('should spawn instrumentation', async () => {
      const userArgs = [];
      const expectedArgs = ['-s', deviceId, 'shell', 'am', 'instrument', '-w', '-r', testRunner];
      await adb.spawnInstrumentation(deviceId, userArgs, testRunner);
      expect(spawn).toHaveBeenCalledWith(adbBinPath, expectedArgs, expect.any(Object));
    });

    it('should pass through additional args', async () => {
      const userArgs = ['-mock', '-args'];
      await adb.spawnInstrumentation(deviceId, userArgs, testRunner);
      expect(spawn).toHaveBeenCalledWith(adbBinPath, expect.arrayContaining([...userArgs, testRunner]), expect.any(Object));
    });

    it('should set detaching=false as spawn-options', async () => {
      const expectedOptions = {
        detached: false,
      };
      const userArgs = [];

      await adb.spawnInstrumentation(deviceId, userArgs, testRunner);

      expect(spawn).toHaveBeenCalledWith(adbBinPath, expect.any(Array), expectedOptions);
    });

    it('should chain-return the promise from spawn util', async () => {
      const userArgs = [];
      const mockPromise = Promise.resolve('mock');
      spawn.mockReturnValue(mockPromise);

      const childProcessPromise = adb.spawnInstrumentation(deviceId, userArgs, testRunner);
      expect(childProcessPromise).toEqual(mockPromise);
    });
  });

  it(`listInstrumentation passes the right deviceId`, async () => {
    const deviceId = 'aDeviceId';
    jest.spyOn(adb, 'shell');

    await adb.listInstrumentation(deviceId);

    expect(adb.shell).toBeCalledWith(deviceId, 'pm list instrumentation');
  });

  it(`getInstrumentationRunner parses the correct runner for the package`, async () => {
    const expectedRunner = 'com.example.android.apis/.app.LocalSampleInstrumentation';
    const expectedPackage = 'com.example.android.apis';
    const instrumentationRunnersShellOutput =
      'instrumentation:com.android.emulator.smoketests/android.support.test.runner.AndroidJUnitRunner (target=com.android.emulator.smoketests)\n' +
      'instrumentation:com.android.smoketest.tests/com.android.smoketest.SmokeTestRunner (target=com.android.smoketest)\n' +
      `instrumentation:${expectedRunner} (target=${expectedPackage})\n` +
      'instrumentation:org.chromium.webview_shell/.WebViewLayoutTestRunner (target=org.chromium.webview_shell)\n';

    jest.spyOn(adb, 'shell').mockImplementation(async () => instrumentationRunnersShellOutput);

    const result = await adb.getInstrumentationRunner('aDeviceId', expectedPackage);

    expect(adb.shell).toBeCalledWith('aDeviceId', 'pm list instrumentation');
    expect(result).toEqual(expectedRunner);
  });

  describe('animation disabling', () => {
    it('should disable animator (e.g. ObjectAnimator) animations', async () => {
      await adb.disableAndroidAnimations();
      expect(exec).toHaveBeenCalledWith(`"${adbBinPath}"  shell "settings put global animator_duration_scale 0"`, { retries: 1 });
    });

    it('should disable window animations', async () => {
      await adb.disableAndroidAnimations();
      expect(exec).toHaveBeenCalledWith(`"${adbBinPath}"  shell "settings put global window_animation_scale 0"`, { retries: 1 });
    });

    it('should disable transition (e.g. activity launch) animations', async () => {
      await adb.disableAndroidAnimations();
      expect(exec).toHaveBeenCalledWith(`"${adbBinPath}"  shell "settings put global transition_animation_scale 0"`, { retries: 1 });
    });
  });
});
