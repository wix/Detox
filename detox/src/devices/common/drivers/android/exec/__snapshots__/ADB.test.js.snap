// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ADB clearAppData should throw an error if the package is not installed 1`] = `
"Failed to clear com.example.app app data on mockEmulator

HINT: Please verify that the package is installed on the device:
adb -s mockEmulator shell pm list packages com.example.app

Command failed"
`;
